src/celeritas/optical/ScintillationGenerator.hh

```diff
@@ -0,0 +1,206 @@
+    inline CELER_FUNCTION Span<OpticalPrimary> operator()(Generator& rng);
+
+    // TODO: Move it to detail/OpticalUtils.
+    static CELER_CONSTEXPR_FUNCTION real_type hc()
+    {
+        return constants::h_planck * constants::c_light;
+    }
+
+  private:
+    //// TYPES ////
+
+    using UniformRealDist = UniformRealDistribution<real_type>;
+    using ExponentialDist = ExponentialDistribution<real_type>;
+
+    //// DATA ////
+
+    OpticalDistributionData const& dist_;
+    ScintillationSpectrum const& spectrum_;
+    Span<OpticalPrimary> photons_;
+
+    UniformRealDist sample_cost_;
+    UniformRealDist sample_phi_;
+    ExponentialDist sample_time_;
+
+    bool is_neutral_{};
+    units::LightSpeed delta_speed_{};
+    Real3 delta_pos_{};
+};
+
+//---------------------------------------------------------------------------//
+// INLINE DEFINITIONS
+//---------------------------------------------------------------------------//
+/*!
+ * Construct from shared scintillation data and distribution parameters.
+ */
+CELER_FUNCTION
+ScintillationGenerator::ScintillationGenerator(
+    OpticalDistributionData const& dist,
+    NativeCRef<ScintillationData> const& shared,
+    Span<OpticalPrimary> photons)
+    : dist_(dist)
+    , spectrum_(shared.spectrum[dist_.material])
+    , photons_(photons)
+    , sample_cost_(-1, 1)
+    , sample_phi_(0, 2 * constants::pi)
+    , sample_time_(real_type{1} / spectrum_.fall_time[0])
+    , is_neutral_{dist_.charge == zero_quantity()}
+{
+    CELER_EXPECT(dist_);
+    CELER_EXPECT(photons_.size() == dist_.num_photons);
+
+    auto const& pre_step = dist_.points[StepPoint::pre];
+    auto const& post_step = dist_.points[StepPoint::post];
+    delta_pos_ = post_step.pos - pre_step.pos;
+    delta_speed_ = post_step.speed - pre_step.speed;
+}
+
+//---------------------------------------------------------------------------//
+/*!
+ * Sample scintillation photons from optical property data and step data.
+ *
+ * The optical photons are generated evenly along the step and are emitted
+ * uniformly over the entire solid angle with a random linear polarization.
+ * The photon energy is calculated by the scintillation emission wavelength
+ * \f[
+   E = \frac{hc}{\lambda},
+ * \f]
+ * where \f$ h \f$ is the Planck constant and \f$ c \f$ is the speed of light,
+ * and \f$ \lambda \f$ is sampled by the normal distribution with the mean of
+ * scintillation emission spectrum and the standard deviation. The emitted time
+ * is simulated according to empirical shapes of the material-dependent
+ * scintillation time structure with one or double exponentials.
+ */
+template<class Generator>
+CELER_FUNCTION Span<OpticalPrimary>
+ScintillationGenerator::operator()(Generator& rng)
+{
+    // Loop for generating scintillation photons
+    size_type num_generated{0};
+
+    for (auto sid : range(spectrum_.size))
+    {
+        // Skip if there is no emission warelength
+        if(spectrum_.lambda_mean[sid] <= 0) 
```

Comment:
When is this the case? Are we sure it corresponds to a zero yield probability? And if so, we shouldn't really need a special "exit early" here since line 145 will be an empty loop because `num_photons = 0`?

Reply:
Okay, this was a temporary solution to protect the zero standard inside `NormalDistribution` (at  [CELER_EXPECT(stddev > 0)](https://github.com/celeritas-project/celeritas/blob/develop/src/celeritas/random/distribution/NormalDistribution.hh#L73); ).  An alternative way is to support an arbitrary number of scintillation components (i.e, not as Real3, but a variable size of array along with num_components.  Any good solution or suggestion?

Reply:
I think we can loosen the assertion to `stddev >= 0`. It may not be "smart" to construct (or sample from) a zero-width gaussian but I don't think it should be prohibited. (It's good to have a "use case" of that edge case, though.)

Reply:
In principle, the requirement is fully legitimated as stddev should be non-zero to define a Gaussian even though the normal distribution will just return the mean for the case.  Not sure whether the relaxing the assertion is a good practice even though it may be not harmful.  The above code looks ugly, but should be removed when we improve the code with an arbitrary number of components (which I do not know how, but should doable if the required memory for the size of SintillationSpectrum (which has all non-zero entries) is pre-allocated on the fly - I need your help for this).

Reply:
Thank you for the tip.  Will try to implement as suggested.

Reply:
@amandalund I modified the ScintillationParams/Data to support an arbitrary number of scintillation components.  Thanks again for the tip!